### PR TITLE
Add craft and office to import and add/reorg AmenityType

### PR DESCRIPTION
**Pull request by tnederlof**
1. Added "craft" and "office" to the importer although not many of these are that relevant and so not many are included in AmenityType but good to have the option to start using these for other usecases in the future.
2. Added a bunch of amenity/shop tags that were not in the existing categories, tried to make a few extra groupings that make sense or from a research/consumer perspective I might want to use.
3.  Alphabatized everything to keep it organized.

I have a bunch of other ideas on how to improve the 15 minute tool but wanted to take a stab at this one first.

### CODE REVIEW

File: map_model/src/objects/building.rs
Comment (michaelkirk) [inline map_model/src/objects/building.rs:231]: Thank you for alphabetizing these =)

File: map_model/src/objects/building.rs
Comment (michaelkirk) [inline map_model/src/objects/building.rs:257]: How do you envision we might use the BikeParkRent amenity?

It's not clear to me that someone would want one when looking for the other.

File: map_model/src/objects/building.rs
Comment (dabreegster) [inline map_model/src/objects/building.rs:430]: It feels like we're outgrowing manually writing these functions to go to/from a string. Glancing around, https://docs.rs/strum/0.20.0/strum/derive.EnumString.html is a popular crate for auto-deriving this with macros. I'm generally hesitant to pile on more dependencies and have the build time/size creep up, but it looks like we're already pulling in the heavier proc macro crates, so I think it'd be worth trying this. The `strum` docs are a little hard to follow, and the crate does lots of stuff we don't seem to need, but https://github.com/Peternator7/strum/blob/master/strum_tests/tests/from_str.rs seems simple to start with

File: map_model/src/objects/building.rs
Comment (dabreegster) [inline map_model/src/objects/building.rs:257]: `amenity=bicycle_parking` is often defined on individual nodes outside of buildings to tag bike racks. With this included here, right now it'll pick up bike parking tagged inside buildings. I think that's totally fine, just wanted to point out that this might be a little unexpected.

File: map_model/src/objects/building.rs
Comment (michaelkirk) [inline map_model/src/objects/building.rs:325]: Cool to see us handling more tags. :+1:

Some of these seem quite obscure - there's such a long tail of tags in OSM, that I don't think adding every single one makes sense, so I'm curious how you're choosing which ones to add?

**Review by michaelkirk — COMMENTED** [review]
👋  Hi @tnederlof - nice to see your PR's coming in.

**Review by dabreegster — COMMENTED** [review]
Nice! Is it fun to play Plato and classify the free-form OSM schema? :)

So normally the process for changing code in the map import pipeline is a little complicated. I would run `data/regen.sh` on my machine, which takes over an hour to rebuild all of the cities configured. I managed to melt my main laptop yesterday (-cough- possibly from running this import process too much and burning out the SSD with writes), and it'll be a few days before I could try to run this again. I'm also looking into moving this process onto AWS.

I'm fine merging the new `craft` and `office` stuff in the meantime, but I won't be able to strenuously test it like normal for a few days.

File: map_model/src/objects/building.rs
Comment (tnederlof) [reply map_model/src/objects/building.rs:257]: When I looked at OSM in certain cities I noticed some bike shares (like citibike in NYC) were being tagged as bike parking. However, I now realize after @dabreegster comment below that these are only buildings being tagged in the 15min section, so maybe better to remove altogether.

File: map_model/src/objects/building.rs
Comment (tnederlof) [reply map_model/src/objects/building.rs:325]: Yes I certainly noticed the long tail. Some of these could be trimmed out and not make much difference, I was trying to go on the side of more is better. There is probably no good way to pick and choose unless maybe we do a specialty shop category to put more of the bespoke types of retailers?

**Comment by tnederlof** [comment]
> 
> 
> Nice! Is it fun to play Plato and classify the free-form OSM schema? :)
> 
> So normally the process for changing code in the map import pipeline is a little complicated. I would run `data/regen.sh` on my machine, which takes over an hour to rebuild all of the cities configured. I managed to melt my main laptop yesterday (-cough- possibly from running this import process too much and burning out the SSD with writes), and it'll be a few days before I could try to run this again. I'm also looking into moving this process onto AWS.
> 
> I'm fine merging the new `craft` and `office` stuff in the meantime, but I won't be able to strenuously test it like normal for a few days.

I barely found any "craft" and "office" categories that made sense to map into the AmenityType categories right now. We could hold off on those two categories for now if you want to wait until you are back up and running to test, totally your call.

File: map_model/src/objects/building.rs
Comment (dabreegster) [reply map_model/src/objects/building.rs:325]: @matkoniecz has helpfully shared https://taginfo.openstreetmap.org/keys/amenity#values before, which exactly shows frequency of use. I think including more is fine for now. We could also consider making a catch-all `Other` category. It depends what we want to do with these categories. The immediate consequence to breaking these down into finer groups is increasing the column height in the 15m tool, and presenting the user with much more nuance up-front.

File: map_model/src/objects/building.rs
Comment (michaelkirk) [reply map_model/src/objects/building.rs:257]: Ah, I see. Thanks for explaining.

**Comment by dabreegster** [comment]
> We could hold off on those two categories for now

I spotted some around Seattle, and it made previously residential-looking buildings appear commercial or mixed-use, so I feel like that's important to capture. I should be able to regenerate everything before the next release, and it's fine to have a little drift accumulate in the meantime.

File: map_model/src/objects/building.rs
Comment (michaelkirk) [reply map_model/src/objects/building.rs:325]: Just in case you weren't already aware of the possibility, one methodical way would be to prioritize based on frequency of tag use — e.g. all the values associated with "shop=*":
https://taginfo.openstreetmap.org/keys/shop#values

File: map_model/src/objects/building.rs
Comment (tnederlof) [reply map_model/src/objects/building.rs:430]: Totally agree, as I was ordering them and copying things around it was quite a pain to keep repeating things. I will take a stab with strum.

File: map_model/src/objects/building.rs
Comment (michaelkirk) [reply map_model/src/objects/building.rs:325]: (whoops, didn't see @dabreegster's comment when I posted)... same same.

File: map_model/src/objects/building.rs
Comment (tnederlof) [reply map_model/src/objects/building.rs:257]: Yes good point, I kind of forgot this is for buildings. I do think things like bike parking, car charging, etc could be useful beyond just the building categories looking forward but certainly plenty to work with using buildings categorized :) I will just remove that whole category I think @michaelkirk also commented on it.

File: map_model/src/objects/building.rs
Comment (tnederlof) [reply map_model/src/objects/building.rs:325]: Looking at the frequency of shop and amenity it does look like a threshold could work so we are somewhat systematic about ignoring the tails. Keeping anything > 0.05% would still keep a lot of the specialty shops but cut out the really obscure ones. Shopping would still be quite a big category though but maybe that's not so terrible.

**Comment by matkoniecz** [comment]
> I barely found any "craft" and "office" categories that made sense to map into the AmenityType categories right now.

I would expect `office=company` `estate_agent` `lawyer` `insurance` `association` `ngo` diplomatic` (embassy, diplomatic mission, consulate or liaison office) to be useful if not supported already.

`craft=carpenter` `shoemaker` `tailor` `photographer` `plumber` seems useful

I have not found this used in code for now, but maybe I missed something? I vaguely remember `office=company` being supported in proletariat robot traffic model - but maybe it was just planed and not done.

**Comment by tnederlof** [comment]
> > I barely found any "craft" and "office" categories that made sense to map into the AmenityType categories right now.
> 
> I would expect `office=company` `estate_agent` `lawyer` `insurance` `association` `ngo` diplomatic` (embassy, diplomatic mission, consulate or liaison office) to be useful if not supported already. https://taginfo.openstreetmap.org/keys/office#values
> 
> `craft=carpenter` `shoemaker` `tailor` `photographer` `plumber` seems useful https://taginfo.openstreetmap.org/keys/craft#values
> 
> I have not found this used in code for now, but maybe I missed something? I vaguely remember `office=company` being supported in proletariat robot traffic model - but maybe it was just planed and not done.

I have to admit I was thinking of this all purely in the context of use in the 15min tool (seeing what buildings are classified as certain AmenityType). For that purpose there were not a ton of craft/office tags that make sense but the ones you mentioned could definilty be helpful.

@dabreegster when you mention commercial or mixed-use is that a different usecase (classifying building useage)? Should AmenityType cover more tags so it can used in different areas of the code and just use different subsets for different usecases (in the 15min tool it would just display a list of AmenityType that someone might reasonably query).

**Comment by dabreegster** [comment]
> I vaguely remember office=company being supported in proletariat robot traffic model

Looks like we never get around to it

> I have to admit I was thinking of this all purely in the context of use in the 15min tool

That's fine for now. The building amenities also get used for [travel demand generation](https://a-b-street.github.io/docs//trafficsim/travel_demand.html), but none of that is fancy yet. Eventually it might recognize that a supermarket category should have more workers and visitors than a small woodworking shop, but not anytime soon.

>  when you mention commercial or mixed-use is that a different usecase (classifying building useage)

I guess that is the 3rd use of this, yeah. The buildings in A/B Street and the 15m tool and everything else are colored either as purely residential or as containing some kind of commercial entity too (either mixed-use or exclusively commercial). So if it represents any sort of business, we ought to have at least one `Amenity` filled out for it, so it appears that way.

**Comment by tnederlof** [comment]
Confusingly I found the tourism key which seems to contain the following keys...

gallery, museum, zoo, hotel, hostel

Looking at the existing 4 keys were are parsing, the gallery and museum tags are not used very much so we should consider tourism.

**Comment by matkoniecz** [comment]
There is also  https://taginfo.openstreetmap.org/keys/leisure#values

=sports_centre =swimming_pool =fitness_centre =marina =stadium =golf_course =resort =water_park =beach_resort

**Comment by dabreegster** [comment]
Mateusz, do you happen to know if there are any existing projects that either attempt to categorize all of the OSM values into broader groups (like our `AmenityType` enum), or even just identify commercial POIs?

**Comment by matkoniecz** [comment]
https://wiki.openstreetmap.org/wiki/SymbolsTab

https://wiki.openstreetmap.org/wiki/Map_features

https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/AddOpeningHours.kt#L22 and below

OsmAnd source code and MAPS.ME source code should be available

iD presets

Vespucci presets

**Comment by tnederlof** [comment]
> 
> 
> https://wiki.openstreetmap.org/wiki/SymbolsTab
> 
> https://wiki.openstreetmap.org/wiki/Map_features
> 
> https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/AddOpeningHours.kt#L22 and below
> 
> OsmAnd source code and MAPS.ME source code should be available
> 
> iD presets
> 
> Vespucci presets

That is really helpful thank you. Looking through that last link it seems like the keys they are using would be the same as ours if we added "leisure" and "tourism" I will add those in my next push to this branch.

**Comment by tnederlof** [comment]
I did a big recategorization and added the main keys into the importer. I also took a stab at replacing the manual Enum into string and string into Enum code that was in buildings.rs, however this is the first time I have delt with adding a dependency to a project this big and using macros in rust so please check it out and yell at me if I did something wrong @dabreegster 

One thing my IDE is yelling at me for are the following (fifteen_min/src/viewer.rs)
```
Error:(335, 41) `&AmenityType` doesn't implement `Display` (required by {})
Error:(336, 62) `&AmenityType` doesn't implement `Display` (required by {})
Error:(491, 54) `AmenityType` doesn't implement `Display` (required by {})
```
Following the examples at: https://docs.rs/strum/0.20.0/strum/derive.Display.html
It looks like I put Display above the AmenityType Enum (which I did) and it should be implemented, however I am getting those messages so I feel like I may me missing something.

**Comment by dabreegster** [comment]
This is looking good! Adding more comments momentarily.

> Error:(335, 41) `&AmenityType` doesn't implement `Display`

The line in question is: `.text(format!("{}: {}", amenity, buildings.len()))`. `amenity` is indeed an `&AmenityType`, a read-only borrow, because we're iterating over a list. The [Display trait](https://doc.rust-lang.org/std/fmt/trait.Display.html) is defined for `&self`, not `self`, so that's why this is valid Rust. And indeed the crate compiles. So basically, your IDE is somehow wrong. I'm not familiar with the Rust IDE landscape, except to know that there's both RLS and rust-analyzer as options. I think it's fine to ignore this warning for now.

File: fifteen_min/src/viewer.rs
Comment (dabreegster) [inline fifteen_min/src/viewer.rs:21]: Style nit: Place `std` imports in their own block before the `abstuil` one here. Unfortunately `cargo fmt` doesn't group imports, so we do this manually. We switched to this "multiple paragraph" style a while ago because somebody was contributing and using an IDE that insisted on this sorting. The order is stuff from `std`, then external crates, then crates within the abstreet repo, then stuff from within the current crate.

File: fifteen_min/Cargo.toml
Comment (dabreegster) [inline fifteen_min/Cargo.toml:24]: You can remove the dependencies here. None of the code in this crate directly uses `strum`; it's all within `map_model`

File: game/src/layer/map.rs
Comment (dabreegster) [inline game/src/layer/map.rs:253]: This actually crashes on the montlake Seattle map now, because one of the categories doesn't match up to a color, and `ColorDiscrete` blindly assumes there's a color defined for every string category you pass in. We can just rewrite this entire function:

```
    pub fn amenities(ctx: &mut EventCtx, app: &App) -> Static {
        let food = Color::RED;
        let school = Color::CYAN;
        let shopping = Color::PURPLE;
        let other = Color::GREEN;

        let mut unzoomed = GeomBatch::new();
        let mut zoomed = GeomBatch::new();
        for b in app.primary.map.all_buildings() {
            if b.amenities.is_empty() {
                continue;
            }
            let mut color = None;
            for a in &b.amenities {
                if let Some(t) = AmenityType::categorize(&a.amenity_type) {
                    color = Some(match t {
                        AmenityType::Food => food,
                        AmenityType::School => school,
                        AmenityType::Shopping => shopping,
                        _ => other,
                    });
                    break;
                }
            }
            let color = color.unwrap_or(other);
            unzoomed.push(color, b.polygon.clone());
            zoomed.push(color.alpha(0.4), b.polygon.clone());
        }

        let panel = Panel::new(Widget::col(vec![
            header(ctx, "Amenities"),
            ColorLegend::row(ctx, food, AmenityType::Food.to_string()),
            ColorLegend::row(ctx, school, AmenityType::School.to_string()),
            ColorLegend::row(ctx, shopping, AmenityType::Shopping.to_string()),
            ColorLegend::row(ctx, other, "other".to_string()),
        ]))
        .aligned_pair(PANEL_PLACEMENT)
        .build(ctx);

        Static {
            panel,
            unzoomed: ctx.upload(unzoomed),
            zoomed: ctx.upload(zoomed),
            name: "amenities",
        }
    }
```
(I really need to learn how to push a commit onto somebody else's PR...)

This rewrite just assigns a color to a few "important"?? categories, and leaves all others as a generic color. This layer can be accessed by running A/B Street (`game` crate), opening layers (hotkey `l`), then amenities (hotkey `a`). I don't know if many people use this layer often; I occasionally open it to quickly understand where businesses are clustered. Picking a different color for every category isn't a useful or maintainable exercise, and this layer never attempted to solve the problem of one building hosting multiple types of shop. If you have better ideas what to do with it, we could try them later, but I don't think this layer is important at all.

File: map_model/src/objects/building.rs
Comment (dabreegster) [inline map_model/src/objects/building.rs:12]: Move these up to line 5 as part of the `serde` block of external crates

File: map_model/src/objects/building.rs
Comment (dabreegster) [inline map_model/src/objects/building.rs:299]: Re-run `cargo +nightly fmt` please; when I did locally, some stuff here changed a bit

**Review by dabreegster — CHANGES_REQUESTED** [review]
Awesome work! The switch to `strum` looks like it's worth it.


When I ran this code locally, my `Cargo.lock` file changed in git. We keep this under version control since abst and the 15m tool are applications, not just libraries. Is your git configured to ignore this somehow?

File: fifteen_min/src/viewer.rs
Comment (tnederlof) [reply fifteen_min/src/viewer.rs:21]: Make sense and yea I am using the Jetbrains IDE where I think they created their own analyzer so its probably just an issue there. I can compile the crate as well so I will ignore it, just wanted to check on it.

File: game/src/layer/map.rs
Comment (tnederlof) [reply game/src/layer/map.rs:253]: Thank you for the rewrite and the explantation, I had not played with that layer yet, it makes sense now.

**Comment by dabreegster** [comment]
The code looks great, the amenity layer is working, and the categories are showing up in the 15m tool, so -- awesome! Thanks for the PR, merging.

**Pull request merged**
